<a href="https://colab.research.google.com/github/rodrigotorresp/Practica-Regresi-n-ML/blob/main/Pr%C3%A1ctica_Regresi%C3%B3n_Entrega.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# IMPORTACIÓN DATOS
Importamos la librería pandsas para guardar y tratar el fichero train.csv

In [27]:
import pandas as pd
train_dataframe= pd.read_csv("https://raw.githubusercontent.com/rodrigotorresp/Practica-Regresi-n-ML/main/train.csv");
train_dataframe

LaptopId  Company                            Product            TypeName  \
0         981  Toshiba                  Portege Z30-C-1CV            Notebook   
1         996       HP                  Spectre 13-V111dx           Ultrabook   
2          77     Dell                      Inspiron 5570            Notebook   
3         103       HP                     Envy 13-ad009n           Ultrabook   
4         946     Dell                      Latitude 7280           Ultrabook   
..        ...      ...                                ...                 ...   
907       994     Dell                      Inspiron 5378  2 in 1 Convertible   
908       615     Asus  FX753VD-GC007T (i7-7700HQ/8GB/1TB              Gaming   
909      1063     Dell                      Inspiron 5567            Notebook   
910        96     Dell                      Inspiron 3567            Notebook   
911       539   Lenovo                      Thinkpad E470            Notebook   

     Inches                           ScreenResolution  \
0      13.3                          Full HD 1920x1080   
1      13.3  IPS Panel Full HD / Touchscreen 1920x1080   
2      15.6                          Full HD 1920x1080   
3      13.3                IPS Panel Full HD 1920x1080   
4      12.5            Full HD / Touchscreen 1920x1080   
..      ...                                        ...   
907    13.3            Full HD / Touchscreen 1920x1080   
908    17.3                          Full HD 1920x1080   
909    15.6                          Full HD 1920x1080   
910    15.6                          Full HD 1920x1080   
911    14.0                IPS Panel Full HD 1920x1080   

                             Cpu   Ram                Memory  \
0     Intel Core i5 6200U 2.3GHz   4GB             128GB SSD   
1     Intel Core i7 7500U 2.7GHz   8GB             256GB SSD   
2     Intel Core i7 8550U 1.8GHz   8GB  128GB SSD +  1TB HDD   
3     Intel Core i7 7500U 2.7GHz   8GB             256GB SSD   
4     Intel Core i5 7200U 2.5GHz   8GB             256GB SSD   
..                           ...   ...                   ...   
907   Intel Core i5 7200U 2.5GHz   8GB             256GB SSD   
908  Intel Core i7 7700HQ 2.8GHz   8GB  128GB SSD +  1TB HDD   
909   Intel Core i7 7500U 2.7GHz  16GB               2TB HDD   
910   Intel Core i7 7500U 2.7GHz   8GB               1TB HDD   
911   Intel Core i7 7500U 2.7GHz   8GB             256GB SSD   

                         Gpu       OpSys  Weight    Price  
0      Intel HD Graphics 520   Windows 7   1.2kg  1195.00  
1      Intel HD Graphics 620  Windows 10  1.11kg  1349.00  
2     Intel UHD Graphics 620  Windows 10  2.02kg   855.00  
3       Nvidia GeForce MX150  Windows 10  1.38kg  1119.00  
4          Intel HD Graphics  Windows 10  1.36kg  1472.20  
..                       ...         ...     ...      ...  
907    Intel HD Graphics 620  Windows 10  1.68kg   889.00  
908  Nvidia GeForce GTX 1050  Windows 10     3kg  1168.00  
909       AMD Radeon R7 M445  Windows 10  2.32kg   989.99  
910       AMD Radeon R5 M430       Linux   2.2kg   599.90  
911     Nvidia GeForce 940MX  Windows 10  1.87kg   859.00  

[912 rows x 13 columns]

# PREPROCESAMIENTO

Uno de los problemas que existen con el tratamiento de estos datos es la existencia de las unidades de medida en algunas columnas, es decir, cuando hablamos de peso el string "kg" está presente en todos los valores del dataset, de manera que es imprescindible eliminarlo de primeras.

De la misma forma, con la columna CPU intentamos separar las unidades de medida utilizando el replace, sustituyendo los espacios en blanco por guiones. Nuestro siguiente paso sería dividir con el método split, en columnas lo que ha sido separado por guiones.

Nuestra intención era utilizar la medida de la frecuencia en GHz, pero al quedarse en algunos casos en columnas distintas no nos fue posible. Sin embargo, pudimos utilizar la marca de procesadores y hacer con ello posteriormente un ordinal encoder, pero no mejoró la predicción.

In [28]:
train_dataframe['Weight'] = train_dataframe['Weight'].str.replace('kg$', '')

train_dataframe['CPU2'] = train_dataframe['Cpu'].str.replace('\s', '-')
cpu = train_dataframe['CPU2'].str.split('[-]',expand=True)

train_dataframe=pd.concat([train_dataframe, cpu], axis=1)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: FutureWarning: The default value of regex will change from True to False in a future version.
  This is separate from the ipykernel package so we can avoid doing imports until


In [29]:
train_dataframe

LaptopId  Company                            Product            TypeName  \
0         981  Toshiba                  Portege Z30-C-1CV            Notebook   
1         996       HP                  Spectre 13-V111dx           Ultrabook   
2          77     Dell                      Inspiron 5570            Notebook   
3         103       HP                     Envy 13-ad009n           Ultrabook   
4         946     Dell                      Latitude 7280           Ultrabook   
..        ...      ...                                ...                 ...   
907       994     Dell                      Inspiron 5378  2 in 1 Convertible   
908       615     Asus  FX753VD-GC007T (i7-7700HQ/8GB/1TB              Gaming   
909      1063     Dell                      Inspiron 5567            Notebook   
910        96     Dell                      Inspiron 3567            Notebook   
911       539   Lenovo                      Thinkpad E470            Notebook   

     Inches                           ScreenResolution  \
0      13.3                          Full HD 1920x1080   
1      13.3  IPS Panel Full HD / Touchscreen 1920x1080   
2      15.6                          Full HD 1920x1080   
3      13.3                IPS Panel Full HD 1920x1080   
4      12.5            Full HD / Touchscreen 1920x1080   
..      ...                                        ...   
907    13.3            Full HD / Touchscreen 1920x1080   
908    17.3                          Full HD 1920x1080   
909    15.6                          Full HD 1920x1080   
910    15.6                          Full HD 1920x1080   
911    14.0                IPS Panel Full HD 1920x1080   

                             Cpu   Ram                Memory  \
0     Intel Core i5 6200U 2.3GHz   4GB             128GB SSD   
1     Intel Core i7 7500U 2.7GHz   8GB             256GB SSD   
2     Intel Core i7 8550U 1.8GHz   8GB  128GB SSD +  1TB HDD   
3     Intel Core i7 7500U 2.7GHz   8GB             256GB SSD   
4     Intel Core i5 7200U 2.5GHz   8GB             256GB SSD   
..                           ...   ...                   ...   
907   Intel Core i5 7200U 2.5GHz   8GB             256GB SSD   
908  Intel Core i7 7700HQ 2.8GHz   8GB  128GB SSD +  1TB HDD   
909   Intel Core i7 7500U 2.7GHz  16GB               2TB HDD   
910   Intel Core i7 7500U 2.7GHz   8GB               1TB HDD   
911   Intel Core i7 7500U 2.7GHz   8GB             256GB SSD   

                         Gpu       OpSys Weight    Price  \
0      Intel HD Graphics 520   Windows 7    1.2  1195.00   
1      Intel HD Graphics 620  Windows 10   1.11  1349.00   
2     Intel UHD Graphics 620  Windows 10   2.02   855.00   
3       Nvidia GeForce MX150  Windows 10   1.38  1119.00   
4          Intel HD Graphics  Windows 10   1.36  1472.20   
..                       ...         ...    ...      ...   
907    Intel HD Graphics 620  Windows 10   1.68   889.00   
908  Nvidia GeForce GTX 1050  Windows 10      3  1168.00   
909       AMD Radeon R7 M445  Windows 10   2.32   989.99   
910       AMD Radeon R5 M430       Linux    2.2   599.90   
911     Nvidia GeForce 940MX  Windows 10   1.87   859.00   

                            CPU2      0     1   2       3       4     5  
0     Intel-Core-i5-6200U-2.3GHz  Intel  Core  i5   6200U  2.3GHz  None  
1     Intel-Core-i7-7500U-2.7GHz  Intel  Core  i7   7500U  2.7GHz  None  
2     Intel-Core-i7-8550U-1.8GHz  Intel  Core  i7   8550U  1.8GHz  None  
3     Intel-Core-i7-7500U-2.7GHz  Intel  Core  i7   7500U  2.7GHz  None  
4     Intel-Core-i5-7200U-2.5GHz  Intel  Core  i5   7200U  2.5GHz  None  
..                           ...    ...   ...  ..     ...     ...   ...  
907   Intel-Core-i5-7200U-2.5GHz  Intel  Core  i5   7200U  2.5GHz  None  
908  Intel-Core-i7-7700HQ-2.8GHz  Intel  Core  i7  7700HQ  2.8GHz  None  
909   Intel-Core-i7-7500U-2.7GHz  Intel  Core  i7   7500U  2.7GHz  None  
910   Intel-Core-i7-7500U-2.7GHz  Intel  Core  i7   7500U  2.7GHz  None  
911   Intel-Core-i7-7500U-2.7GH

  - Eliminamos LaptopId ya que no nos aporta ningun valor el tener el identificador del portátil.

  - Podríamos haber tenido en cuenta la compañía que fabrica el dispositvo pero revisando los datos, hay gran varianza en los precios de las diferentes compañías.

  - La columna 'Product' decidimos prescindir de ella, ya que considerabamos que no aportaba información.

  - Con la columna 'Typename' realimos un ordinalencoder, dándole más importancia a 'Workstation' que a las demás. Posteriormente aplicaremos también un StandardScaler, para ajustar los datos y que el modelo pueda aprender bien.

  - Con respecto a 'Inches', en la primera transformación pasamos de largo, pero en la segunda aplicamos también un StandardScaler.

  - 'ScreenResolution' se intentó proceder de la misma manera que con Cpu (antes comentado) y al no conseguir el valor numérico de la resolución en una misma columna fue imposible utilizar está información.

  - Eliminamos la columna Cpu, utilizaremos las columnas que concatenamos en el anterior bloque de código.

  - Con la memoria RAM utilizamos un OrdinalEncoder por categorías, yendo de mayor a menor prioridad desde 2GB hasta 64GB.

  - Tuvimos la idea de utilizar la información de la columna 'Memory' de manera que tuviesemos una columna que nos indicase con un 1 o un 0 si la memoria era de tipo HDD o de tipo SSD, utilizando la sentencia OneHotEncoder, pero no fuimos capaces de aislar en una misma columna el string 'HDD' o 'SSD' para utilizarlo. Al final la eliminamos

  - Con la columna 'GPU' nos encontramos con el mismo problema.

  - Utilizamos de nuevo el OrdinalEncoder con categorías para ordenar por prioridad el sistema operativo. Posteriormente aplicaremos StandardScaler.

  - En el caso de 'Weight' no realiazamos ningun cambio en el primer bloque de código, para posteriormente usar StandardScaler.

  - Eliminamos 'Price' porque es nuestra variable objetivo, por tanto no debe existir en el conjunto de entrenamiento.

  - 'CPU2' se elimina porque es el vector separado por guiones que utilizamos para separar por columnas.

  - 'INTEL/AMD' sería la columna que usaríamos para hacer un ordinalEnconder con Intel AMD y Samsung, pero decidimos no usarlo ya que no mejoraba las predicciones.

  - Las demás columnas se eliminaron puesto que no pudimos asilar en una misma columna los datos necesarios para usar esa información (antes explicado)

In [30]:
import sklearn.compose
colum_transformer = sklearn.compose.ColumnTransformer(transformers=[
    ("Laptop Id" , "drop", [0]),
    ("Company" , "drop", [1]),
    ("Product" , "drop", [2]),
    ("TypeName" , sklearn.preprocessing.OrdinalEncoder(categories=[['Netbook', 'Notebook', 'Ultrabook','Gaming','2 in 1 Convertible','Workstation']]), [3]),
    ("Inches" , "passthrough", [4]),
    ("ScreenResolution" , "drop", [5]),
    ("CPU" , "drop", [6]),
    ("RAM" , sklearn.preprocessing.OrdinalEncoder(categories=[['2GB', '4GB','6GB','8GB','12GB','16GB','24GB','32GB' ,'64GB']]), [7]),
    ("Memory" , "drop", [8]),
    ("GPU" , "drop", [9]),
    ("OPSYS" , sklearn.preprocessing.OrdinalEncoder(categories=[['No OS', 'Android' , 'Chrome OS' , 'Windows 7','Linux','Windows 10','macOS','Windows 10 S','Mac OS X']]), [10]),
    ("Weight" , "passthrough", [11]),
    ("Price" , "drop", [12]),
    ("CPU2", "drop", [13]),
    ("INTEL/AMD", "drop", [14]),
    ("1", "drop", [15]),
    ("2", "drop", [16]),
    ("3", "drop", [17]),
    ("4", "drop", [18]),
    ("5", "drop", [19]),
    ]);

train_dataframe_transform = colum_transformer.fit_transform(train_dataframe)
train_dataframe_transform

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:1692: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['int', 'str']. An error will be raised in 1.2.
  FutureWarning,


array([[1.0, 13.3, 1.0, 3.0, '1.2'],
       [2.0, 13.3, 3.0, 5.0, '1.11'],
       [1.0, 15.6, 3.0, 5.0, '2.02'],
       ...,
       [1.0, 15.6, 5.0, 5.0, '2.32'],
       [1.0, 15.6, 3.0, 4.0, '2.2'],
       [1.0, 14.0, 3.0, 5.0, '1.87']], dtype=object)

In [31]:
import sklearn.compose
colum_transformer_Scaler = sklearn.compose.ColumnTransformer(transformers=[
    ("TypeNameScaler" , sklearn.preprocessing.StandardScaler(), [0]),
    ("InchesScaler" , sklearn.preprocessing.StandardScaler(), [1]),
    ("RAMScaler" , sklearn.preprocessing.StandardScaler(), [2]),
    ("WeightScaler" , sklearn.preprocessing.StandardScaler(), [4]),
    ("OPSYSScaler" , sklearn.preprocessing.StandardScaler(), [3]),
    ]);

train_dataframe_transform = colum_transformer_Scaler.fit_transform(train_dataframe_transform)
train_dataframe_transform

array([[-0.71119323, -1.20960244, -1.19779584, -1.26029362, -1.3613048 ],
       [ 0.15131771, -1.20960244,  0.17455316, -1.39490179,  0.30291769],
       [-0.71119323,  0.40762114,  0.17455316, -0.0338636 ,  0.30291769],
       ...,
       [-0.71119323,  0.40762114,  1.54690217,  0.41483031,  0.30291769],
       [-0.71119323,  0.40762114,  0.17455316,  0.23535275, -0.52919356],
       [-0.71119323, -0.71740396,  0.17455316, -0.25821055,  0.30291769]])

En primera instancia, pensamos en una regresión lineal, ya que no teníamos muchos parámetros de entrada y parecía viable a primera vista. 

In [32]:
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
 
X = train_dataframe_transform.data
y = train_dataframe.Price

X_train = train_dataframe_transform.data
y_train = train_dataframe.Price

reg = LinearRegression().fit(X, y)
reg.intercept_

1130.0370065789473

Su valor es 1130.0370065789473 que representa el precio básico de cualquier portátil sin contar con el resto de las características que describan los portátiles.

Ya habiendo entrenado la regresión lineal 'reg' revisamos sus métricas de evaluacíon

In [33]:
from sklearn.metrics import mean_absolute_error, mean_squared_error

y_pred = reg.predict(X_train)
print("MAE: ", mean_absolute_error(y_train, y_pred))
print("MSE: ", mean_squared_error(y_train, y_pred))
print("R2: ", reg.score(X_train, y_train))

MAE:  315.8897097253008
MSE:  195928.9706337869
R2:  0.605863683358332


Hicimos una primera submission y decidimos probar otros modelos para comparar y ver si alguno lo mejoraba

Usamos Lasso porque sospechabamos que varios de los atributos de entrada serían irrelevantes, pero no generó buenos resultados.

In [34]:
from sklearn.linear_model import Lasso

lasso = Lasso(alpha=0).fit(X_train,y_train)
y_pred = lasso.predict(X_train)
print("MAE: ", mean_absolute_error(y_train, y_pred))
print("MSE: ", mean_squared_error(y_train, y_pred))
print("R2: ", lasso.score(X_train, y_train))

lasso2 = Lasso(alpha=2.5).fit(X_train,y_train)
y_pred = lasso.predict(X_train)
print("MAE: ", mean_absolute_error(y_train, y_pred))
print("MSE: ", mean_squared_error(y_train, y_pred))
print("R2: ", lasso2.score(X_train, y_train))

lasso5 = Lasso(alpha=5.0).fit(X_train,y_train)
y_pred = lasso.predict(X_train)
print("MAE: ", mean_absolute_error(y_train, y_pred))
print("MSE: ", mean_squared_error(y_train, y_pred))
print("R2: ", lasso5.score(X_train, y_train))

lasso7 = Lasso(alpha=7.5).fit(X_train,y_train)
y_pred = lasso.predict(X_train)
print("MAE: ", mean_absolute_error(y_train, y_pred))
print("MSE: ", mean_squared_error(y_train, y_pred))
print("R2: ", lasso7.score(X_train, y_train))

MAE:  315.8897097253008
MSE:  195928.9706337869
R2:  0.605863683358332
MAE:  315.8897097253008
MSE:  195928.9706337869
R2:  0.6057001776126527
MAE:  315.8897097253008
MSE:  195928.9706337869
R2:  0.6052129241219247
MAE:  315.8897097253008
MSE:  195928.9706337869
R2:  0.6044006124638815


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  This is separate from the ipykernel package so we can avoid doing imports until
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_coordinate_descent.py:648: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  coef_, l1_reg, l2_reg, X, y, max_iter, tol, rng, random, positive
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_coordinate_descent.py:648: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 8.934e+07, tolerance: 4.534e+04 Linear regression models with null weight for the l1 regularization term are more efficiently fitted using one of the solvers implemented in sklearn.linear_model.Ridge

Dado que Lasso dió malos resultados,optamos por probar con una regresión lineal pero con regularización, como lo es Ridge. Ya que requiere que los atributos este correlados entre si.Aplicamos distintos valores al hiper-parámetro alpha para ver cual producía el menor error.

In [36]:
from sklearn.linear_model import Ridge

ridge1 = Ridge(alpha=1.0).fit(X_train, y_train)
y_pred = ridge1.predict(X_train)
print("MAE: ", mean_absolute_error(y_train, y_pred))
print("MSE: ", mean_squared_error(y_train, y_pred))
print("R2: ", ridge1.score(X_train, y_train))

ridge5 = Ridge(alpha=5.0).fit(X_train, y_train)
y_pred = ridge5.predict(X_train)
print("MAE: ", mean_absolute_error(y_train, y_pred))
print("MSE: ", mean_squared_error(y_train, y_pred))
print("R2: ", ridge5.score(X_train, y_train))

ridge10 = Ridge(alpha=10.0).fit(X_train, y_train)
y_pred = ridge10.predict(X_train)
print("MAE: ", mean_absolute_error(y_train, y_pred))
print("MSE: ", mean_squared_error(y_train, y_pred))
print("R2: ", ridge10.score(X_train, y_train))

ridge15 = Ridge(alpha=15.0).fit(X_train, y_train)
y_pred = ridge15.predict(X_train)
print("MAE: ", mean_absolute_error(y_train, y_pred))
print("MSE: ", mean_squared_error(y_train, y_pred))
print("R2: ", ridge15.score(X_train, y_train))

MAE:  315.8507760742898
MSE:  195929.34592678654
R2:  0.6058629284082087
MAE:  315.70685841077477
MSE:  195938.1661438636
R2:  0.6058451854074696
MAE:  315.53868172438337
MSE:  195964.8735551151
R2:  0.6057914600157439
MAE:  315.4164438504895
MSE:  196007.89723956832
R2:  0.6057049123425546


Tembién probamos la regresión de Bayes pero no generó mejores resultados

In [37]:
from sklearn import linear_model

Bayes = linear_model.BayesianRidge()
Bayes.fit(X_train, y_train)
y_pred = Bayes.predict(X_train)

print("MAE: ", mean_absolute_error(y_train, y_pred))
print("MSE: ", mean_squared_error(y_train, y_pred))
print("R2: ", Bayes.score(X_train, y_train))

MAE:  315.7464493663559
MSE:  195934.5437146503
R2:  0.6058524723895979


Usamos la regresión tree, que es buen modelo a la hora de predecir una variable a partir de variables independientes entre si, pero resulto generar overfitting

In [38]:
from sklearn import tree
tree = tree.DecisionTreeRegressor()
tree = tree.fit(X_train, y_train)
y_pred = tree.predict(X_train)

print("MAE: ", mean_absolute_error(y_train, y_pred))
print("MSE: ", mean_squared_error(y_train, y_pred))
print("R2: ", tree.score(X_train, y_train))

MAE:  88.7247180837599
MSE:  21782.98972875928
R2:  0.9561807153410525


Al usar la regresión KNeighbors dió mejores resultados
Los parametros significan:
- n_neighbors: Número de vecinos que se usarán de forma predeterminada para las consultas de kneighbors.
- weights: Ponderaciones. Al usar 'uniform' todos los puntos en cada vecindario se ponderan por igual.

In [39]:
from sklearn.neighbors import KNeighborsRegressor

knn = KNeighborsRegressor(n_neighbors=10, weights='uniform').fit(X_train, y_train)
y_pred = knn.predict(X_train)
print("MAE: ", mean_absolute_error(y_train, y_pred))
print("MSE: ", mean_squared_error(y_train, y_pred))
print("R2: ", knn.score(X_train, y_train))

MAE:  240.83543092105262
MSE:  134993.96331331908
R2:  0.7284422855024332


In [44]:
test_dataframe= pd.read_csv("https://raw.githubusercontent.com/rodrigotorresp/Practica-Regresi-n-ML/main/test.csv");
test_dataframe;
test_dataframe['Weight'] = test_dataframe['Weight'].str.replace('kg$', '')

test_dataframe['CPU2'] = test_dataframe['Cpu'].str.replace('\s', '-')
cpu = test_dataframe['CPU2'].str.split('[-]',expand=True)

test_dataframe=pd.concat([test_dataframe, cpu], axis=1)

import sklearn.compose
colum_transformer = sklearn.compose.ColumnTransformer(transformers=[
    ("Laptop Id" , "drop", [0]),
    ("Company" , "drop", [1]),
    ("Product" , "drop", [2]),
    ("TypeName" , sklearn.preprocessing.OrdinalEncoder(categories=[['Netbook', 'Notebook', 'Ultrabook','Gaming','2 in 1 Convertible','Workstation']]), [3]),
    ("Inches" , "passthrough", [4]),
    ("ScreenResolution" , "drop", [5]),
    ("CPU" , "drop", [6]),
    ("RAM" , sklearn.preprocessing.OrdinalEncoder(categories=[['2GB', '4GB','6GB','8GB','12GB','16GB','24GB','32GB' ,'64GB']]), [7]),
    ("Memory" , "drop", [8]),
    ("GPU" , "drop", [9]),
    ("OPSYS" , sklearn.preprocessing.OrdinalEncoder(categories=[[ 'No OS', 'Android' , 'Chrome OS' , 'Windows 7','Linux','Windows 10','macOS','Windows 10 S','Mac OS X']]), [10]),
    ("Weight" , "passthrough", [11]),
    ("CPU2", "drop", [12]),
    ("INTEL/AMD", "drop", [13]),
    ("1", "drop", [14]),
    ("2", "drop", [15]),
    ("3", "drop", [16]),
    ("4", "drop", [17]),
    ("5", "drop", [18]),
    ]);
  
test_dataframe_transform = colum_transformer.fit_transform(test_dataframe);

colum_transformer_Scaler = sklearn.compose.ColumnTransformer(transformers=[
    ("TypeNameScaler" , sklearn.preprocessing.StandardScaler(), [0]),
    ("InchesScaler" , sklearn.preprocessing.StandardScaler(), [1]),
    ("RAMScaler" , sklearn.preprocessing.StandardScaler(), [2]),
    ("WeightScaler" , sklearn.preprocessing.StandardScaler(), [4]),
    ("OPSYSScaler" , sklearn.preprocessing.StandardScaler(), [3]),
    ]);

test_dataframe_transform = colum_transformer_Scaler.fit_transform(test_dataframe_transform);

y_test = knn.predict(test_dataframe_transform);

compression_opts = dict(method='zip',
                        archive_name='Submission.csv')

submissionDataFrame = pd.DataFrame({"LaptopId": test_dataframe.LaptopId, "price": y_test});
submissionDataFrame.to_csv("Submission.zip", index=False, compression=compression_opts);

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: FutureWarning: The default value of regex will change from True to False in a future version.
  This is separate from the ipykernel package so we can avoid doing imports until
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: FutureWarning: The default value of regex will change from True to False in a future version.
  """
/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:1692: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['int', 'str']. An error will be raised in 1.2.
  FutureWarning,
